centml/compiler/prediction/profiler.py

```diff
@@ -0,0 +1,162 @@
+            flattened_shapes = []
+            dtypes = []
+
+            for arg in args:
+                if isinstance(arg, (tuple, list)):
+                    if len(arg) > 0 and isinstance(arg[0], (tuple, list, torch.Tensor)):
+                        nested_shapes, nested_dtypes = get_flattened_shapes(arg[0])
+                        shape = [len(arg)] + nested_shapes
+                        dtypes.extend(nested_dtypes.split(','))
+                    else:
+                        shape = [len(arg)]
+                elif isinstance(arg, torch.Tensor):
+                    shape = list(arg.shape)
+                    dtypes.append(str(arg.dtype))
+                elif isinstance(arg, bool):
+                    shape = [1 if arg is True else 0]
+                elif isinstance(arg, (int, float)):
+                    shape = [arg]
+                else:
+                    shape = [1]
+                flattened_shapes.extend(shape)
+
+            if len(flattened_shapes) < 2:
+                flattened_shapes.extend([1])
+
+            input_dtypes = ','.join(dtypes) if dtypes else 'N/A'
+
+            return flattened_shapes, input_dtypes
+
+        def get_output_dtypes(results):
+            def find_dtypes(results):
+                if isinstance(results, torch.Tensor):
+                    return [str(results.dtype)]
+                if isinstance(results, (list, tuple)):
+                    dtypes = []
+                    for item in results:
+                        dtypes.extend(find_dtypes(item))
+                    return dtypes
+                return []
+
+            types = find_dtypes(results)
+
+            if types:
+                return ','.join(types)
+            return 'N/A'
+
+        for node in self.graph.nodes:
+            result = None
+            if node.op == 'placeholder':
+                result = next(args_iter)
+            elif node.op == 'get_attr':
+                result = fetch_attr(node.target)
+            elif node.op == 'call_function':
+                args = load_arg(node.args)
+                kwargs = load_arg(node.kwargs)
+                result = node.target(*args, **kwargs)
+
+                inp_shapes, input_dtypes = get_flattened_shapes(args)
+                output_dtypes = get_output_dtypes(result)
+
+                key = (node.target.__name__, len(inp_shapes), input_dtypes, output_dtypes, self.gpu)
+
+                t = self.tree_db.get(key, inp_shapes)
+
+                if self.data_collection_mode and t is None:
+                    with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CUDA]) as prof:
+                        result = node.target(*args, **kwargs)
+                    event_time_total = 0
+                    for event in prof.key_averages():
+                        event_time_total += event.cuda_time_total
+                    t = event_time_total
+                    self.tree_db.add(key, inp_shapes, t)
```

Comment:
can you move this logic to a function as well? This code is repeated 3 times. You can pass in the callable operation  that need to be profiled into the function.